Libraries/LibGfx/ICOLoader.cpp

```diff
@@ -0,0 +1,387 @@
+    size_t data_size { 0 };
+    Vector<ImageDescriptor> images;
+    size_t largest_index;
+};
+
+RefPtr<Gfx::Bitmap> load_ico(const StringView& path)
+{
+    MappedFile mapped_file(path);
+    if (!mapped_file.is_valid())
+        return nullptr;
+    ICOImageDecoderPlugin decoder((const u8*)mapped_file.data(), mapped_file.size());
+    auto bitmap = decoder.bitmap();
+    if (bitmap)
+        bitmap->set_mmap_name(String::format("Gfx::Bitmap [%dx%d] - Decoded ICO: %s", bitmap->width(), bitmap->height(), LexicalPath::canonicalized_path(path).characters()));
+    return bitmap;
+}
+
+RefPtr<Gfx::Bitmap> load_ico_from_memory(const u8* data, size_t length)
+{
+    ICOImageDecoderPlugin decoder(data, length);
+    auto bitmap = decoder.bitmap();
+    if (bitmap)
+        bitmap->set_mmap_name(String::format("Gfx::Bitmap [%dx%d] - Decoded ICO: <memory>", bitmap->width(), bitmap->height()));
+    return bitmap;
+}
+
+static Optional<size_t> decode_ico_header(BufferStream& stream)
+{
+    ICONDIR header;
+    stream.read_raw((uint8_t*)&header, sizeof(header));
+    if (stream.handle_read_failure())
+        return {};
+
+    if (header.must_be_0 != 0 || header.must_be_1 != 1)
+        return {};
+    return { header.image_count };
+}
+
+static Optional<ImageDescriptor> decode_ico_direntry(BufferStream& stream)
+{
+    ICONDIRENTRY entry;
+    stream.read_raw((uint8_t*)&entry, sizeof(entry));
+    if (stream.handle_read_failure())
+        return {};
+
+    ImageDescriptor desc = { entry.width, entry.height, entry.offset, entry.size, nullptr };
+    if (desc.width == 0)
+        desc.width = 256;
+    if (desc.height == 0)
+        desc.height = 256;
+
+    return { desc };
+}
+
+static bool load_ico_directory(ICOLoadingContext& context)
+{
+    auto buffer = ByteBuffer::wrap(context.data, context.data_size);
+    auto stream = BufferStream(buffer);
+    auto image_count = decode_ico_header(stream);
+    if (!image_count.has_value() || image_count.value() == 0) {
+        return false;
+    }
+
+    size_t max_area = 0;
+    for (size_t i = 0; i < image_count.value(); ++i) {
+        auto maybe_desc = decode_ico_direntry(stream);
+        if (!maybe_desc.has_value()) {
+#ifdef ICO_DEBUG
+            printf("load_ico_directory: error loading entry: %lu\n", i);
+#endif
+            return false;
+        }
+
+        auto& desc = maybe_desc.value();
+        // FIXME: Overflow
+        if (desc.offset >= context.data_size
+            || (desc.offset + desc.size) > context.data_size) {
+#ifdef ICO_DEBUG
+            printf("load_ico_directory: offset: %lu size: %lu doesn't fit in ICO size: %lu\n",
+                desc.offset, desc.size, context.data_size);
+#endif
+            return false;
+        }
+#ifdef ICO_DEBUG
+        printf("load_ico_directory: index %lu width: %u height: %u offset: %lu size: %lu\n",
+            i, desc.width, desc.height, desc.offset, desc.size);
+#endif
+        if (desc.width * desc.height > max_area) {
```

Comment:
Personally, I find it a bit surprising that finding the largest-area image is mixed with decoding. I'd probably do this elsewhere.